hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -706,7 +706,20 @@ protected void compactIfNecessary(AbstractHoodieWriteClient writeClient, String
     }
   }
 
-  protected void doClean(AbstractHoodieWriteClient writeClient, String instantTime) {
+  protected void cleanIfNecessary(AbstractHoodieWriteClient writeClient, String instantTime) {
+    Option<HoodieInstant> lastCompletedCompactionInstant = metadataMetaClient.reloadActiveTimeline()
+        .getCommitTimeline().filterCompletedInstants().lastInstant();
+    if (lastCompletedCompactionInstant.isPresent()
+        && metadataMetaClient.getActiveTimeline().filterCompletedInstants()
+            .findInstantsAfter(lastCompletedCompactionInstant.get().getTimestamp()).countInstants() < 3) {
+      // do not clean the log files immediately after compaction to give some buffer time for metadata table reader,
+      // because there is case that the reader has prepared for the log file readers already before the compaction completes
+      // while before/during the reading of the log files, the cleaning triggers and delete the reading files,
+      // then a FileNotFoundException(for LogFormatReader) or NPE(for HFileReader) would throw.
+
+      // 3 is a value that I think is enough for metadata table reader.
```

Comment:
what's this based on?

Reply:
Just a subjective value that i think the metadata table reads fast enough within 3 transaction interval. Do you have better suggestion ?